### PR TITLE
Serve content from memory

### PR DESCRIPTION
**Pull request by carver**
### What was wrong?

May not always want to offer content from database, sometimes want to offer it from memory. See #368 

LRU cache is a hacky solution, this is ... a little less hacky.

### How was it fixed?

Added a new `PopulatedOffer` which is just like an `Offer` but with the data included.

This is a special-cased out-bound only offer that allows us to skip all the database lookups. Even more importantly, we can offer data that *never* gets stored locally, like in the bridge nodes.

There are a couple of commits added at the end, related to the peertest failure in CI. The commits are pretty tidy, for reviewing one at a time.

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [ ] Add entry to the [release notes](https://github.com/ethereum/trin/blob/master/newsfragments/README.md) (may forgo for trivial changes)
- [ ] Clean up commit history
- [x] Add commit that reverts LRU cache

### CODE REVIEW

**Comment by carver** [comment]
Not able to reproduce the peertest failure locally. I assume this is the same thing that everyone else is running into.

**Comment by carver** [comment]
So it looks like we have a little more info with the peertest failure, when the `exit(1)` is removed:
```
Jul 28 16:17:00.840  INFO ethportal_peertest::jsonrpc: Testing IPC method: "portal_historyStore"    
Jul 28 16:17:01.026 DEBUG trin_core::jsonrpc::service: New IPC client: (unnamed)    
Jul 28 16:17:01.028 DEBUG trin_core::jsonrpc::service: Got new IPC request: Ok(JsonRequest { jsonrpc: "2.0", params: Array([String("0x000300720704f3aa11c53cf344ea069db95cecb81ad7453c8f276b2a1062979611f09c"), String("0xf90222a02c58e3212c085178dbb1277e2f3c24b3f451267a75a234945c1581af639f4a7aa058a694212e0416353a4d3865ccf475496b55af3a3d3b002057000741af9731919400192fb10df37c9fb26829eb2cc623cd1bf599e8a067a9fb631f4579f9015ef3c6f1f3830dfa2dc08afe156f750e90022134b9ebf6a018a2978fc62cd1a23e90de920af68c0c3af3330327927cda4c005faccefb5ce7a0168a3827607627e781941dc777737fc4b6beb69a8b139240b881992b35b854eab9010000200000400000001000400080080000000000010004010001000008000000002000110000000000000090020001110402008000080208040010000000a8000000000000000000210822000900205020000000000160020020000400800040000000000042080000000400004008084020001000001004004000001000000000000001000000110000040000010200844040048101000008002000404810082002800000108020000200408008000100000000000000002020000b00010080600902000200000050000400000000000000400000002002101000000a00002000003420000800400000020100002000000000000000c00040000001000000100187327bd7ad3116ce83e147ed8401c9c36483140db184627d9afa9a457468657265756d50504c4e532f326d696e6572735f55534133a0f1a32e24eb62f01ec3f2b3b5893f7be9062fbf5482bc0d490a54352240350e26882087fbb243327696851aae1651b6")]), method: "portal_historyStore", id: 11 })    
Jul 28 16:17:01.028 DEBUG trin_core::jsonrpc::handlers: Processing Portal JSON-RPC request: HistoryEndpoint(Store)    
Jul 28 16:17:01.352 DEBUG discv5::discv5: Discv5 service already shutdown
Jul 28 16:17:01.353 DEBUG discv5::discv5: Discv5 service already shutdown
Jul 28 16:17:01.353 DEBUG trin_core::jsonrpc::service: Panic hook: Skipped removing /tmp/trin-jsonrpc.ipc because: No such file or directory (os error 2)    
Jul 28 16:17:01.353 DEBUG trin_core::jsonrpc::service: Panic hook: Skipped removing /tmp/ethportal-peertest-buddy-2.ipc because: No such file or directory (os error 2)    
Jul 28 16:17:01.353 DEBUG trin_core::jsonrpc::service: Panic hook: Skipped removing /tmp/ethportal-peertest-buddy-1.ipc because: No such file or directory (os error 2)    
Jul 28 16:17:01.561 DEBUG discv5::discv5: Discv5 service already shutdown
test test::test_launches has been running for over 60 seconds
```

There was a panic, but no information printed about the panic :/

**Comment by carver** [comment]
... and now it looks like it's hanging, which is presumably why the `exit(1)` was there in the first place. We just need to find a better solution for a clean rpc exit I suppose.

File: trin-core/src/portalnet/overlay.rs [hunk 1]
```diff
@@ -294,26 +294,24 @@ where
 
         // HashMap to temporarily store all interested ENRs and the content.
         // Key is base64 string of node's ENR.
-        let mut enrs_and_content: HashMap<String, Vec<RawContentKey>> = HashMap::new();
+        let mut enrs_and_content: HashMap<String, Vec<(RawContentKey, ByteList)>> = HashMap::new();
 
         // Filter all nodes from overlay routing table where XOR_distance(content_id, nodeId) < node radius
-        for content_key_value in content {
+        for content_item in content {
```
Comment (njgheorghita) [inline trin-core/src/portalnet/overlay.rs:300]: *nit* `for (content_key, content_value) in content` could get rid of the `0` and `1` indexing further down

**Review by njgheorghita — APPROVED** [review]
Looks good! Agree it's a cleaner strategy than the lru cache. As we start supporting a header accumulator, it seems like a cache might be a tool we want to use... but also maybe it'll be better to store them. Either way, agreed let's get rid of the cache until we find a good use case to bring it back

**Comment by carver** [comment]
Failure on Windows build with:
```Aug 03 01:36:53.828  INFO ethportal_peertest::scenarios: Testing OFFER/ACCEPT flow    
Aug 03 01:36:54.028 DEBUG trin_core::jsonrpc::service: New IPC client: (unnamed)    
Aug 03 01:36:54.031 DEBUG trin_core::jsonrpc::service: Got new IPC request: Ok(JsonRequest { jsonrpc: "2.0", params: Array([String("0x000300720704f3aa11c53cf344ea069db95cecb81ad7453c8f276b2a1062979611f09c"), String("0xf90222a02c58e3212c085178dbb1277e2f3c24b3f451267a75a234945c1581af639f4a7aa058a694212e0416353a4d3865ccf475496b55af3a3d3b002057000741af9731919400192fb10df37c9fb26829eb2cc623cd1bf599e8a067a9fb631f4579f9015ef3c6f1f3830dfa2dc08afe156f750e90022134b9ebf6a018a2978fc62cd1a23e90de920af68c0c3af3330327927cda4c005faccefb5ce7a0168a3827607627e781941dc777737fc4b6beb69a8b139240b881992b35b854eab9010000200000400000001000400080080000000000010004010001000008000000002000110000000000000090020001110402008000080208040010000000a8000000000000000000210822000900205020000000000160020020000400800040000000000042080000000400004008084020001000001004004000001000000000000001000000110000040000010200844040048101000008002000404810082002800000108020000200408008000100000000000000002020000b00010080600902000200000050000400000000000000400000002002101000000a00002000003420000800400000020100002000000000000000c00040000001000000100187327bd7ad3116ce83e147ed8401c9c36483140db184627d9afa9a457468657265756d50504c4e532f326d696e6572735f55534133a0f1a32e24eb62f01ec3f2b3b5893f7be9062fbf5482bc0d490a54352240350e26882087fbb243327696851aae1651b6")]), method: "portal_historyStore", id: 11 })    
Aug 03 01:36:54.031 DEBUG trin_core::jsonrpc::handlers: Processing Portal JSON-RPC request: HistoryEndpoint(Store)    
Aug 03 01:36:54.231 DEBUG trin_core::jsonrpc::service: New IPC client: (unnamed)    
Aug 03 01:36:54.232 DEBUG trin_core::jsonrpc::service: Got new IPC request: Ok(JsonRequest { jsonrpc: "2.0", params: Array([String("enr:-IS4QKjoo7Ks3ghrWSROc6UzZ0jD3ENsVwEKyz6Um1yT8sgKQGP97-jzZod-HTdDtUTKu96_n2nR-VtWR1E7O35oRU4BgmlkgnY0gmlwhH8AAAGJc2VjcDI1NmsxoQLNT3JLMGY9ao9J0CPP6tlcGXyEtLpESGp-UV4La22VAYN1ZHCCIyk"), Array([String("0x000300720704f3aa11c53cf344ea069db95cecb81ad7453c8f276b2a1062979611f09c")])]), method: "portal_historyOffer", id: 11 })    
Aug 03 01:36:54.232 DEBUG trin_core::jsonrpc::handlers: Processing Portal JSON-RPC request: HistoryEndpoint(Offer)    
Aug 03 01:36:54.241 DEBUG discv5::service: Sending RPC Request: id: f90e8d775e2935e2: TALK: protocol: 500b, request: 060400000004000000000300720704f3aa11c53cf344ea069db95cecb81ad7453c8f276b2a1062979611f09c to node: Node: 0x8a27..1b0f, addr: 127.0.0.1:9001
Aug 03 01:36:54.243 DEBUG trin_core::portalnet::overlay_service: [History] Node set to Connected. Node: 0xb871..66a9    
Aug 03 01:36:54.244 DEBUG trin_core::portalnet::overlay_service: [History] Handling request f90e8d775e2935e2    
Aug 03 01:36:54.244 DEBUG trin_history::events: Sending reply: Accept(Accept { connection_id: 40728, content_keys: Bitfield { bytes: [1], len: 1, _phantom: PhantomData } })
Aug 03 01:36:54.245 DEBUG discv5::service: Sending TALK response to Node: 0xb871..66a9, addr: 127.0.0.1:9000
Aug 03 01:36:54.246 DEBUG discv5::service: Received RPC response: Response: Response 07189f0600000003 to request: TALK: protocol: 500b, request: 060400000004000000000300720704f3aa11c53cf344ea069db95cecb81ad7453c8f276b2a1062979611f09c from: Node: 0x8a27..1b0f, addr: 127.0.0.1:9001
Aug 03 01:36:54.247 DEBUG trin_core::portalnet::overlay_service: [History] Node set to Connected. Node: 0x8a27..1b0f    
Aug 03 01:36:54.248 DEBUG trin_core::utp::stream: sent Packet { type: Syn, version: 1, extension: None, connection_id: 6303, timestamp: Timestamp(1150419510), timestamp_difference: Delay(0), wnd_size: 0, seq_nr: 1, ack_nr: 0 }    
Aug 03 01:36:54.249 DEBUG discv5::service: Sending RPC Request: id: 4d0540d6e6430252: TALK: protocol: 757470, request: 4100189f44920236000000000000000000010000 to node: Node: 0x8a27..1b0f, addr: 127.0.0.1:9001
Aug 03 01:36:54.250 DEBUG trin_core::utp::stream: received Packet { type: Syn, version: 1, extension: None, connection_id: 6303, timestamp: Timestamp(1150419510), timestamp_difference: Delay(0), wnd_size: 0, seq_nr: 1, ack_nr: 0 }    
Aug 03 01:36:54.251 DEBUG trin_core::utp::stream: Handle packet: Syn. Conn state: Uninitialized    
Aug 03 01:36:54.251 DEBUG discv5::service: Sending RPC Request: id: 427d8ab1e56bf044: TALK: protocol: 757470, request: 2100189f44920e6100000c2b00100000d3360001 to node: Node: 0xb871..66a9, addr: 127.0.0.1:9000
Aug 03 01:36:54.253 DEBUG discv5::service: Sending empty TALK response to Node: 0x8a27..1b0f, addr: 127.0.0.1:9001
Aug 03 01:36:54.254 DEBUG trin_core::utp::stream: received Packet { type: State, version: 1, extension: None, connection_id: 6303, timestamp: Timestamp(1150422625), timestamp_difference: Delay(3115), wnd_size: 1048576, seq_nr: 54070, ack_nr: 1 }    
Aug 03 01:36:54.254 DEBUG trin_core::utp::stream: Handle packet: State. Conn state: SynSent    
Aug 03 01:36:54.255 DEBUG trin_core::utp::stream: sent Packet { type: Data, version: 1, extension: None, connection_id: 6304, timestamp: Timestamp(1150426547), timestamp_difference: Delay(3388), wnd_size: 0, seq_nr: 2, ack_nr: 54069 }    
Aug 03 01:36:54.254 DEBUG discv5::service: Received RPC response: Response: Response  to request: TALK: protocol: 757470, request: 2100189f44920e6100000c2b00100000d3360001 from: Node: 0xb871..66a9, addr: 127.0.0.1:9000
Aug 03 01:36:54.255 DEBUG trin_core::utp::stream: sent Packet { type: State, version: 1, extension: None, connection_id: 6303, timestamp: Timestamp(1150422625), timestamp_difference: Delay(3115), wnd_size: 1048576, seq_nr: 54070, ack_nr: 1 }    
Aug 03 01:36:54.255 DEBUG discv5::service: Sending empty TALK response to Node: 0xb871..66a9, addr: 127.0.0.1:9000
Aug 03 01:36:54.255 DEBUG discv5::service: Sending RPC Request: id: f8f5aedd8d84ee9a: TALK: protocol: 757470, request: 010018a044921db300000d3c000000000002d33504000000f90222a02c58e3212c085178dbb1277e2f3c24b3f451267a75a234945c1581af639f4a7aa058a694212e0416353a4d3865ccf475496b55af3a3d3b002057000741af9731919400192fb10df37c9fb26829eb2cc623cd1bf599e8a067a9fb631f4579f9015ef3c6f1f3830dfa2dc08afe156f750e90022134b9ebf6a018a2978fc62cd1a23e90de920af68c0c3af3330327927cda4c005faccefb5ce7a0168a3827607627e781941dc777737fc4b6beb69a8b139240b881992b35b854eab9010000200000400000001000400080080000000000010004010001000008000000002000110000000000000090020001110402008000080208040010000000a8000000000000000000210822000900205020000000000160020020000400800040000000000042080000000400004008084020001000001004004000001000000000000001000000110000040000010200844040048101000008002000404810082002800000108020000200408008000100000000000000002020000b00010080600902000200000050000400000000000000400000002002101000000a00002000003420000800400000020100002000000000000000c00040000001000000100187327bd7ad3116ce83e147ed8401c9c36483140db184627d9afa9a457468657265756d50504c4e532f326d696e6572735f55534133a0f1a32e24eb62f01ec3f2b3b5893f7be9062fbf5482bc0d490a54352240350e26882087fbb243327696851aae1651b6 to node: Node: 0x8a27..1b0f, addr: 127.0.0.1:9001
Aug 03 01:36:54.256 DEBUG discv5::service: Received RPC response: Response: Response  to request: TALK: protocol: 757470, request: 4100189f44920236000000000000000000010000 from: Node: 0x8a27..1b0f, addr: 127.0.0.1:9001
Aug 03 01:36:54.258 DEBUG trin_core::utp::stream: received Packet { type: Data, version: 1, extension: None, connection_id: 6304, timestamp: Timestamp(1150426547), timestamp_difference: Delay(3388), wnd_size: 0, seq_nr: 2, ack_nr: 54069 }    
Aug 03 01:36:54.258 DEBUG trin_core::utp::stream: Handle packet: Data. Conn state: Connected    
Aug 03 01:36:54.259 DEBUG discv5::service: Sending RPC Request: id: 28b06fe535dc994a: TALK: protocol: 757470, request: 2100189f44922d6f00000fbc00100000d3360002 to node: Node: 0xb871..66a9, addr: 127.0.0.1:9000
Aug 03 01:36:54.261 DEBUG discv5::service: Sending empty TALK response to Node: 0x8a27..1b0f, addr: 127.0.0.1:9001
Aug 03 01:36:54.262 DEBUG trin_core::utp::stream: received Packet { type: State, version: 1, extension: None, connection_id: 6303, timestamp: Timestamp(1150430575), timestamp_difference: Delay(4028), wnd_size: 1048576, seq_nr: 54070, ack_nr: 2 }    
Aug 03 01:36:54.262 DEBUG trin_core::utp::stream: Handle packet: State. Conn state: Connected    
Aug 03 01:36:54.262 DEBUG trin_core::jsonrpc::service: New IPC client: (unnamed)    
Aug 03 01:36:54.262 DEBUG discv5::service: Received RPC response: Response: Response  to request: TALK: protocol: 757470, request: 2100189f44922d6f00000fbc00100000d3360002 from: Node: 0xb871..66a9, addr: 127.0.0.1:9000
Aug 03 01:36:54.263 DEBUG trin_core::jsonrpc::service: Got new IPC request: Ok(JsonRequest { jsonrpc: "2.0", params: Array([String("0x000300720704f3aa11c53cf344ea069db95cecb81ad7453c8f276b2a1062979611f09c")]), method: "portal_historyLocalContent", id: 16 })    
Aug 03 01:36:54.263 DEBUG discv5::service: Sending RPC Request: id: dc9db414e522a67b: TALK: protocol: 757470, request: 110018a044923b9300000000000000000003d335 to node: Node: 0x8a27..1b0f, addr: 127.0.0.1:9001
Aug 03 01:36:54.263 DEBUG trin_core::utp::stream: sent Packet { type: State, version: 1, extension: None, connection_id: 6303, timestamp: Timestamp(1150430575), timestamp_difference: Delay(4028), wnd_size: 1048576, seq_nr: 54070, ack_nr: 2 }    
Aug 03 01:36:54.263 DEBUG trin_core::utp::stream: Removed packet from incoming buffer: Packet { type: Data, version: 1, extension: None, connection_id: 6304, timestamp: Timestamp(1150426547), timestamp_difference: Delay(3388), wnd_size: 0, seq_nr: 2, ack_nr: 54069 }    
Aug 03 01:36:54.264 DEBUG trin_core::jsonrpc::handlers: Processing Portal JSON-RPC request: HistoryEndpoint(LocalContent)    
Aug 03 01:36:54.264 DEBUG discv5::service: Sending empty TALK response to Node: 0xb871..66a9, addr: 127.0.0.1:9000
Aug 03 01:36:54.266 DEBUG trin_core::utp::stream: received Packet { type: Fin, version: 1, extension: None, connection_id: 6304, timestamp: Timestamp(1150434195), timestamp_difference: Delay(0), wnd_size: 0, seq_nr: 3, ack_nr: 54069 }    
Aug 03 01:36:54.266 DEBUG trin_core::utp::stream: Handle packet: Fin. Conn state: Connected    
Aug 03 01:36:54.266 DEBUG trin_core::utp::stream: FIN received but there are missing acknowledgements for sent packets    
Aug 03 01:36:54.266 DEBUG discv5::service: Received RPC response: Response: Response  to request: TALK: protocol: 757470, request: 010018a044921db300000d3c000000000002d33504000000f90222a02c58e3212c085178dbb1277e2f3c24b3f451267a75a234945c1581af639f4a7aa058a694212e0416353a4d3865ccf475496b55af3a3d3b002057000741af9731919400192fb10df37c9fb26829eb2cc623cd1bf599e8a067a9fb631f4579f9015ef3c6f1f3830dfa2dc08afe156f750e90022134b9ebf6a018a2978fc62cd1a23e90de920af68c0c3af3330327927cda4c005faccefb5ce7a0168a3827607627e781941dc777737fc4b6beb69a8b139240b881992b35b854eab9010000200000400000001000400080080000000000010004010001000008000000002000110000000000000090020001110402008000080208040010000000a8000000000000000000210822000900205020000000000160020020000400800040000000000042080000000400004008084020001000001004004000001000000000000001000000110000040000010200844040048101000008002000404810082002800000108020000200408008000100000000000000002020000b00010080600902000200000050000400000000000000400000002002101000000a00002000003420000800400000020100002000000000000000c00040000001000000100187327bd7ad3116ce83e147ed8401c9c36483140db184627d9afa9a457468657265756d50504c4e532f326d696e6572735f55534133a0f1a32e24eb62f01ec3f2b3b5893f7be9062fbf5482bc0d490a54352240350e26882087fbb243327696851aae1651b6 from: Node: 0x8a27..1b0f, addr: 127.0.0.1:9001
Aug 03 01:36:54.267 DEBUG trin_core::jsonrpc::service: Panic hook: Skipped removing /tmp/trin-jsonrpc.ipc because: The system cannot find the file specified. (os error 2)    
Aug 03 01:36:54.268 DEBUG trin_core::jsonrpc::service: Panic hook: Skipped removing /tmp/ethportal-peertest-buddy-2.ipc because: The system cannot find the file specified. (os error 2)    
Aug 03 01:36:54.268 DEBUG trin_core::jsonrpc::service: Panic hook: Skipped removing /tmp/ethportal-peertest-buddy-1.ipc because: The system cannot find the file specified. (os error 2)    
Aug 03 01:36:54.268 DEBUG discv5::service: Sending empty TALK response to Node: 0xb871..66a9, addr: 127.0.0.1:9000
Aug 03 01:36:54.268  WARN discv5::service: Failed to send empty talk response channel closed
Aug 03 01:36:54.269 DEBUG discv5::discv5: Discv5 service already shutdown
Aug 03 01:36:54.270 DEBUG discv5::discv5: Discv5 service already shutdown
Aug 03 01:36:54.332 DEBUG discv5::discv5: Discv5 service already shutdown
test test::test_launches has been running for over 60 seconds

Too long with no output (exceeded 10m0s): context deadline exceeded
```

**Pull request merged**